clarity/src/vm/functions/principals.rs

```diff
@@ -38,13 +69,259 @@ pub fn special_is_standard(
+    // This is an aborting error because this should have been caught in analysis pass.
+    let verified_hash_bytes = match hash_bytes {
+        Value::Sequence(SequenceData::Buffer(BuffData { ref data })) => data,
+        _ => return Err(CheckErrors::TypeValueError(BUFF_20.clone(), hash_bytes).into()),
+    };
+
+    // This must have been a (buff 20).
+    // This is an aborting error because this should have been caught in analysis pass.
+    if verified_hash_bytes.len() > 20 {
+        return Err(CheckErrors::TypeValueError(BUFF_20.clone(), hash_bytes).into());
+    }
+
+    // If the hash-bytes buffer has less than 20 bytes, this is a runtime error, because it
+    // wasn't the job of the type system (i.e. (buff X) for all X < 20 are all also (buff 20))
+    if verified_hash_bytes.len() < 20 {
+        return Ok(create_principal_true_error_response(1));
+    }
+
+    // Construct the principal.
+    let mut transfer_buffer = [0u8; 20];
+    transfer_buffer.copy_from_slice(&verified_hash_bytes);
+    let principal_data = StandardPrincipalData(version_byte, transfer_buffer);
+
+    let principal = if let Some(name) = name_opt {
+        // requested a contract principal.  Verify that the `name` is a valid ContractName.
+        // The type-checker will have verified that it's (string-ascii 40), but not long enough.
+        let name_bytes = match name {
+            Value::Sequence(SequenceData::String(CharType::ASCII(ref ascii_data))) => {
+                &ascii_data.data
+            }
+            _ => {
+                return Err(CheckErrors::TypeValueError(
+                    TypeSignature::contract_name_string_ascii(),
+                    name,
+                )
+                .into())
+            }
+        };
+
+        // If it's not long enough, then it's a runtime error that warrants an (err ..) response.
+        if name_bytes.len() < CONTRACT_MIN_NAME_LENGTH {
+            return Ok(create_principal_true_error_response(2));
+        }
+
+        // if it's too long, then this should have been caught by the type-checker
+        if name_bytes.len() > CONTRACT_MAX_NAME_LENGTH {
+            return Err(CheckErrors::TypeValueError(
+                TypeSignature::contract_name_string_ascii(),
+                name,
+            )
+            .into());
+        }
+
+        // The type-checker can't verify that the name is a valid ContractName, so we'll need to do
+        // it here at runtime.  If it's not valid, then it warrants this function evaluating to
+        // (err ..).
+        let name_string = match name {
+            // destruct again to avoid a .clone() on the inner ascii_data.data
+            Value::Sequence(SequenceData::String(CharType::ASCII(ascii_data))) => {
+                String::from_utf8(ascii_data.data).expect("FAIL: could not convert bytes of type (string-ascii 40) back to a UTF-8 string")
+            },
+            _ => {
+                unreachable!()
+            }
+        };
```

Comment:
This code can be cleaned up so that it doesn't do a second destruct while still avoiding a clone, though it adds a "restructuring" in an unreachable error path (it shouldn't be possible in production to get that `TypeValue` error).

```suggestion
        let name_bytes = match name {
            Value::Sequence(SequenceData::String(CharType::ASCII(ascii_data))) => {
                ascii_data
            }
            _ => {
                return Err(CheckErrors::TypeValueError(
                    TypeSignature::contract_name_string_ascii(),
                    name,
                )
                .into())
            }
        };

        // If it's not long enough, then it's a runtime error that warrants an (err ..) response.
        if name_bytes.len() < CONTRACT_MIN_NAME_LENGTH {
            return Ok(create_principal_true_error_response(2));
        }

        // if it's too long, then this should have been caught by the type-checker
        if name_bytes.len() > CONTRACT_MAX_NAME_LENGTH {
            return Err(CheckErrors::TypeValueError(
                TypeSignature::contract_name_string_ascii(),
                Value::from(name_bytes),
            )
            .into());
        }

        // The type-checker can't verify that the name is a valid ContractName, so we'll need to do
        // it here at runtime.  If it's not valid, then it warrants this function evaluating to
        // (err ..).
        let name_string = String::from_utf8(name_bytes.data).expect("FAIL: could not convert bytes of type (string-ascii 40) back to a UTF-8 string");
```

Reply:
Fixed